packages/core/src/serialize.ts

```diff
@@ -0,0 +1,191 @@
+     *  Serialize published properties to JSON Object
+     *  Usage:
+     *      const obj = serialize(someWidgetInstance);
+     */
+    serialize(): SerializationObject;
+
+    /**
+     *  Deserialize JSON Object into class
+     *  Usage:
+     *      deserialize(someWidgetInstance, obj);
+     */
+    deserialize(_: SerializationObject): void;
+
+    /**
+     *  Check if a published property has a value
+     *  @param prop: propery ID
+     */
+    exists<K extends keyof this>(prop: K): boolean;
+
+    /**
+     *  Check if a published property has had a value set by the user
+     *  @param prop: propery ID
+     */
+    modified<K extends keyof this>(prop: K): boolean;
+
+    /**
+     *  Reset the published property back to its default value
+     *  @param prop: propery ID
+     */
+    reset<K extends keyof this>(prop: K): void;
+}
+
+export type SerializationObject = {
+    module: string,
+    class: string,
+    properties: { [key: string]: any }
+};
+
+/**
+ *  Annotation to enable serialization and publish properties on any given class
+ *  Usage:
+ *      @serializable("chart_Column")
+ *      class Column extends LiteWidget {
+ *          ...
+ *      }
+ */
+export function serializable(moduleName: string) {
+    // tslint:disable-next-line: callable-types
+    return function <T extends { new(...args: any[]): {} }>(constructor: T) {
+        const __proto__ = constructor.prototype;
+        const className = __proto__.constructor.name;
+
+        return class extends constructor {
+
+            _messageScheduler = new MessageSchedular();
+
+            className(): string {
+                return className;
+            }
+
+            cssClass(prefix: string = "hpcc-js"): string {
+                if (__proto__.cssClass) {
+                    return __proto__.cssClass.call(this) + " " + prefix + "-" + className;
+                }
+                return prefix + "-" + className;
+            }
+
+            serialize(): SerializationObject {
+                const retVal = {
+                    module: moduleName,
+                    class: className,
+                    properties: {} as any
+                };
+                for (const key in __proto__) {
+                    if (is_META(key)) {
+                        const property = from_META(key);
+                        if ((this as any).modified(property)) {
+                            retVal.properties[property] = (this as any)[property]();
+                        }
+                    }
+                }
+                return retVal;
+            }
+
+            deserialize(_: SerializationObject) {
+                if (moduleName !== _.module || className !== _.class) throw new Error("Module / Class does not match");
+                for (const key in __proto__) {
+                    if (is_META(key)) {
+                        const property = from_META(key);
+                        (this as any)[property](_.properties[property]);
```

Comment:
This is goes along with my suggestion for `serialize`. This line should become something like:
```javascript
let val = _.properties[property];
if(val && val.module){
    val = new Widget().deserialize(val);
}
(this as any)[property](val);
```